[Feng1231] ip

Feng1231 opened the pull request:
## Duck
> "A plan is what; a schedule is when. It takes both a plan and a schedule to get things done." - Peter Turla

Are you:

- [x] Having too many daily tasks to keep track
- [x] Having upcoming deadlines
- [x] In search for a task scheduler

If you tick any of those, then you need our ~~QUACK~~ **Duck**!

#### Duck allows you to schedule without a hassle. It's
- Friendly, Quack! 🦆
- Easy to get your hands on
- No distraction elements e.g. advertisements

Simply:
1. download it [here](https://github.com/Feng1231/ip/releases/download/A-Release/duck-0.2-all.jar)
2. double click on it.

Here are the available commands for Duck.
```java
public enum CommandWord {
    /** Command to list all tasks. */
    LIST,

    /** Command to find tasks by a keyword. */
    FIND,

    /** Command to mark a task as done. */
    MARK,

    /** Command to unmark a task as incomplete. */
    UNMARK,

    /** Command to delete a task. */
    DELETE,

    /** Command to add a ToDo task. */
    TODO,

    /** Command to add a Deadline task. */
    DEADLINE,

    /**  Command to add an Event task. */
    EVENT,

    /** Command to list tasks due on a specific date. */
    ON,

    /** Command to exit the application. */
    BYE
}
```
and that's it! You're ready to get organised!

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Feng1231 committed:
Rename, Greet, Exit

Feng1231 committed:
Echo user input

Feng1231 committed:
Add, List

Feng1231 committed:
Mark as Done

Feng1231 committed:
ToDo, Event, Deadline: Add 3 types of tasks

Feng1231 committed:
text-ui-test: Add test cases

Feng1231 committed:
DukeException: Handle Exceptions

Feng1231 committed:
Enable delete and use enum class for instructions

Feng1231 committed:
update code to meet expected output

Feng1231 committed:
improve code style

Feng1231 committed:
Rename DukeException class to DuckException, create data folder to store data

Feng1231 committed:
Task and its subclasses: Add toDataString() method, Duck: Allow create and read from file

Feng1231 committed:
Add methods to allow append/write to dataFile

Feng1231 committed:
Task and its subclass: Add new constructor that can specify isDone attribute

Feng1231 committed:
Make read task from file create task with correct isDone status

Feng1231 committed:
Task and its subclasses: Change constructor method signature

Feng1231 committed:
Duck.java: Handle case for corrupted data file, Add print message to inform starting up of Duck

Feng1231 committed:
Merge branch 'branch-Level-7'

Feng1231 committed:
Duck.java: Add function to convert valid date-time string into LocalDateTime

Feng1231 committed:
Allow LocalDateTime to be saved in data file in correct format

Feng1231 committed:
Event.java: Enforce end is after start, Duck.java: Allow search for task with 'on' instruction

Feng1231 committed:
Refactor Duck to include more OOP, package application

Feng1231 committed:
Merge branch 'add-gradle-support'

Feng1231 committed:
Add JUnit Tests, Set Up build.gradle configs

Feng1231 committed:
Remove commented out code in Duck

Feng1231 committed:
Add JavaDocs to all the files

Feng1231 committed:
Adhere to coding standards

Feng1231 committed:
Add find functionality to Duck

Feng1231 committed:
Merge branch 'branch-A-CodingStandard'

Feng1231 committed:
Merge branch-Level-9 and fix conflict

Feng1231 committed:
Add config files for checkstyle

Feng1231 committed:
Follow code style in all files

Feng1231 committed:
Merge branch 'branch-A-CheckStyle'

Feng1231 committed:
Add gitignore to ignore data folder

Feng1231 committed:
Add Nodes for JavaFX, Add handleUserInput method in Duck

Feng1231 committed:
Add start up message, Exit application upon Bye command

Feng1231 committed:
Enable render warning message when data file has invalid format

Feng1231 committed:
Add HelpCommand to navigate user through the commands

Feng1231 committed:
Update UI of nodes

Feng1231 committed:
Add AI.md, change images, update to HelpCommand HELP_STRING

Feng1231 committed:
Update build gradle archive version, Update Duck.java to pass stylecheck

Feng1231 committed:
Update data file path to absolute path

Currently, the data file path is relative to the project root directory.
This causes the file to be created in the root directory of the project.
This has caused problem in the JAR file as the file destination mightnot be found.

Let's update the data file path to an absolute path to prevent this issue.
The storage test cases also have been tweaked to use absolute path for testing.

yong-zhuo commented on src/test/java/duck/storage/StorageTest.java:
```diff
@@ -0,0 +1,164 @@
+     * @throws DuckException If there is an issue with the Duck-specific exception handling.
+     * @throws IOException   If there is an issue with file operations.
+     */
+    @RepeatedTest(3)
+    public void writeTasks_repeatedly_shouldOverWrite() throws DuckException, IOException {
```
Your test cases are quite extensive, and I like that you've chosen to repeat the tests multiple times to ensure consistent results. Great job!

yong-zhuo replied:
Your test cases are quite extensive, and I like that you've chosen to repeat the tests multiple times to ensure consistent results. Great job!

yong-zhuo commented on src/main/java/duck/ui/Ui.java:
```diff
@@ -0,0 +1,105 @@
+    /**
+     * Displays the greeting message along with an ASCII logo and a prompt for user interaction.
+     */
+    public void sayHi() {
```
Nice ASCII logo for your bot :)

yong-zhuo commented on src/main/java/duck/parser/Parser.java:
```diff
@@ -0,0 +1,63 @@
+        try {
+            CommandWord instruction = CommandWord.valueOf(getInstruction(message));
+            //CHECKSTYLE.OFF: Indentation
+            return switch (instruction) {
+                case HELP -> new HelpCommand(message);
+                case LIST -> new ListCommand(message);
+                case FIND -> new FindCommand(message);
+                case MARK -> new MarkCommand(message);
+                case UNMARK -> new UnmarkCommand(message);
+                case DELETE -> new DeleteCommand(message);
+                case TODO -> new ToDoCommand(message);
+                case DEADLINE -> new DeadlineCommand(message);
+                case EVENT -> new EventCommand(message);
+                case ON -> new OnCommand(message);
+                case BYE -> new ByeCommand(message);
```
I like how well-structured this is and how it follows the OOP principles 👍

yong-zhuo commented on src/main/java/duck/Duck.java:
```diff
@@ -0,0 +1,139 @@
+        });
+    }
+
+    private String captureOutput(Runnable runnable) {
+        // Create a ByteArrayOutputStream to capture the output
+        ByteArrayOutputStream baos = new ByteArrayOutputStream();
+
+
+        PrintStream originalOut = System.out; // Save the original System.out
+        System.setOut(new PrintStream(baos)); // Redirect System.out to the ByteArrayOutputStream
+
+        runnable.run();
+
+        // Ensure all output is flushed
+        System.out.flush();
+
+        // Convert the captured output to a string
+        return baos.toString();
+    }
```
This is a good and interesting approach to capturing the output. Nicely done!

Feng1231 replied:
> This is a good and interesting approach to capturing the output. Nicely done!

Thanks! I thought this used extraction method to improve code readability.

yong-zhuo reviewed (commented):
Overall LGTM! Variables are well named and code follows the coding standards. OOP principles are followed and I like how you structured your code and designed your test cases. Very well done! 👍

Feng1231 committed:
Add assert statements to the codebase

The methods are invoked with non null assumptions made on the parameters.

The arguments passed to the methods could be null due to implementation errors.
This could lead to unwanted behaviour in the program, such as exception being thrown.

Let's add assert statements to the codebase to ensure that the assumptions made on the parameters are met.

Feng1231 committed:
Refactor codebase to improve code quality

Magic literals are used many where in the codebase.
This makes the codebase hard to read and maintain.

Let's introduce constants to replace magic literals in the codebase.

Using constants will improve readability and maintainability of the codebase.

Feng1231 committed:
Rename package duck.util to duck.common

Utils class, containing common utility functions used across various classes was packaged with util package.

Let's:
- Rename the package to duck.common
- Create a Message class to store all the common messages used across the application

This allows the duck.common package to contain all common classes and functions used across the application.

Feng1231 committed:
Refactor more classes to improve code quality

Feng1231 committed:
Refactor more classes to improve code quality

Feng1231 committed:
Fix bug after regression testing

Feng1231 committed:
Update javadocs to reflect code changes

Feng1231 committed:
Merge pull request #1 from Feng1231/branch-A-Assertions

Add assert statements to the codebase

Feng1231 committed:
Resolve merge conflicts with master branch

Feng1231 committed:
Update build.gradle to enable assertion

Feng1231 committed:
Merge pull request #2 from Feng1231/branch-A-CodeQuality

Refactor code base to improve code quality

Feng1231 committed:
Mesasge.java: Update javadoc

Feng1231 committed:
Merge pull request #3 from Feng1231/branch-A-Streams

Mesasge.java: Update javadoc

Feng1231 committed:
Update constants to change Duck's responses

Feng1231 committed:
gradle.yml: Add workflow with GitHub Actions

Feng1231 committed:
build.gradle: Update shadowJar archiveVersion

Feng1231 committed:
Update build.gradle

Feng1231 committed:
Update build.gradle showdowJar Version

Feng1231 committed:
Add Sort command

Feng1231 committed:
Refactor code to improve code quality

Feng1231 committed:
Refactor Sort Command and Add parameter package

There are a lot of constants used in the `SortCommand` class.
This reduce the readability and maintainability of the code.

Let's
- Add a parameter package
- Create a `By` enum class under parameter package to represent relevant `By` parameters
- Create a `Target` enum class under parameter package to represent relavant `Target` parameters
- Refactor the `SortCommand` class to refer to the two enum classes instead of using defined constants.

This will make the code more readable and extendable.

Feng1231 committed:
Alter checkstyle.xml

Feng1231 committed:
Alter checkstyle.fxml and update import order

Feng1231 committed:
Update error message for sort command

Feng1231 committed:
Merge pull request #4 from Feng1231/branch-BCD-Extension

Implement Sort Command

Feng1231 committed:
Add test classes

Feng1231 committed:
Fix stylecheck issues

Feng1231 committed:
Update javadocs, Update Invalid Command execute method

Feng1231 committed:
Add more automated testing

Feng1231 committed:
Handle error of having extra spaces between words

Feng1231 committed:
Add javadocs for test methods

Feng1231 committed:
Add app title and icon

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update Ui.png

Feng1231 committed:
Rename png

Feng1231 committed:
Rename png

Feng1231 committed:
Update Ui.png

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update README.md in docs folder

Feng1231 committed:
Update build.gradle shadowJar archiveVersion

Feng1231 committed:
Fix javadocs coding standard

Feng1231 committed:
Update javadocs

Feng1231 committed:
Update docs/README.md

Feng1231 committed:
Update docs/README.md
